### PR TITLE
feat(etl): :sparkles: implement collect/upload logic(not flush)

### PR DESCRIPTION
**Pull request by Eason0729**
When a `Delivery` arrive:
1. Parse to Heartbeat. (deserialize)
2. Tranform to `BeatBuffer`. (Place data to `trie`)
3. Put in `BeatsBuffer`. (Sorted by Uuid and detect active event that cause buffer to be ready for uploading)

image

There are further work to be done:
1. Implement a `struct` for passive event(timeout) that cause buffer to be ready for uploading.
2. Implement a merging logic.

[sync-focus-PR-Model.zip](https://github.com/gdsc-ncku/sync-focus/files/15074057/sync-focus-PR-Model.zip)

### CODE REVIEW

**Comment by peterxcli** [comment]
@Eason0729 Remember to add a duration service for calculating `Vec<Heartbeat>`
reference: [duration.py](https://github.com/gdsc-ncku/sync-focus/blob/main/service/duration.py)

**Comment by peterxcli** [comment]
@Eason0729 LGTM, if you want to add the duration service at next PR, you can close this manually. Thx!

File: etl/src/process/trie.rs [hunk 1]
```diff
@@ -0,0 +1,92 @@
+//! prefix tree
+
+use std::{collections::BTreeMap, ops::Deref};
+
+#[derive(Default)]
```
Comment (peterxcli) [inline etl/src/process/trie.rs:5]: Isn't there any library that has implemented Trie?

**Review by peterxcli — APPROVED** [review]
Nice Work!

File: etl/src/process/trie.rs (hunk 1 above)
Comment (Eason0729) [reply etl/src/process/trie.rs:5]: Yes, but that's not what we needed.
What `trie_rs` does is **only** add count at leaf node, but we need it along path.
For example, we have `http://example.com/about/eason` and `http://example.com/about`, the user just see the fact he visitied `http://example.com/about` twice.

Comment (peterxcli) [reply etl/src/process/trie.rs:5]: Great, thank you for your explanation ~

**Pull request merged**